### ISSUE TITLE
persistent-mysql test failure with MariaDB

### ISSUE THREAD

**Issue by felixonmars** [yesodweb/persistent#1158]
* Your OS name and version
Arch Linux x86_64
* The versions of tools you're using (e.g. `stack`, `ghc`).
GHC 8.10.2
* The versions of dependencies you're using
MariaDB 10.5.8

persistent-mysql is at 2.10.2.3 (didn't test the latest version yet).

```
Running 1 test suites...
Test suite test: RUNNING...
Migrating: CREATe TABLE `Person`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`age` BIGINT NOT NULL,`color` TEXT CHARACTER SET utf8mb4 NULL)
Migrating: CREATe TABLE `Person1`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`age` BIGINT NOT NULL)
Migrating: CREATe TABLE `PersonMaybeAge`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`age` BIGINT NULL)
Migrating: CREATe TABLE `PersonMay`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`name` TEXT CHARACTER SET utf8mb4 NULL,`color` TEXT CHARACTER SET utf8mb4 NULL)
Migrating: CREATe TABLE `Pet`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`ownerId` BIGINT NOT NULL REFERENCES `Person`,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`type` TEXT CHARACTER SET utf8mb4 NOT NULL)
CREATe TABLE `Person`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`age` BIGINT NOT NULL,`color` TEXT CHARACTER SET utf8mb4 NULL);
ALTER TABLE `Person` ADD CONSTRAINT `PersonNameKey` UNIQUE(`name`(200));
CREATe TABLE `Person1`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`age` BIGINT NOT NULL);
CREATe TABLE `PersonMaybeAge`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`age` BIGINT NULL);
CREATe TABLE `PersonMay`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`name` TEXT CHARACTER SET utf8mb4 NULL,`color` TEXT CHARACTER SET utf8mb4 NULL);
CREATe TABLE `Pet`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`ownerId` BIGINT NOT NULL REFERENCES `Person`,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`type` TEXT CHARACTER SET utf8mb4 NOT NULL);
ALTER TABLE `Pet` ADD CONSTRAINT `Pet_ownerId_fkey` FOREIGN KEY(`ownerId`) REFERENCES `Person`(`id`);
CREATe TABLE `MaybeOwnedPet`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`ownerId` BIGINT NULL REFERENCES `Person`,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`type` TEXT CHARACTER SET utf8mb4 NOT NULL);
ALTER TABLE `MaybeOwnedPet` ADD CONSTRAINT `MaybeOwnedPet_ownerId_fkey` FOREIGN KEY(`ownerId`) REFERENCES `Person`(`id`);
CREATe TABLE `NeedsPet`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`petKey` BIGINT NOT NULL REFERENCES `Pet`);
ALTER TABLE `NeedsPet` ADD CONSTRAINT `NeedsPet_petKey_fkey` FOREIGN KEY(`petKey`) REFERENCES `Pet`(`id`);
CREATe TABLE `OutdoorPet`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`ownerId` BIGINT NOT NULL REFERENCES `Person`,`collar` TEXT CHARACTER SET utf8mb4 NOT NULL,`type` TEXT CHARACTER SET utf8mb4 NOT NULL);
ALTER TABLE `OutdoorPet` ADD CONSTRAINT `OutdoorPet_ownerId_fkey` FOREIGN KEY(`ownerId`) REFERENCES `Person`(`id`);
CREATe TABLE `UserPT`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`ident` TEXT CHARACTER SET utf8mb4 NOT NULL,`password` TEXT CHARACTER SET utf8mb4 NULL);
ALTER TABLE `UserPT` ADD CONSTRAINT `UniqueUserPT` UNIQUE(`ident`(200));
CREATe TABLE `EmailPT`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`email` TEXT CHARACTER SET utf8mb4 NOT NULL,`user` BIGINT NULL REFERENCES `UserPT`,`verkey` TEXT CHARACTER SET utf8mb4 NULL);
ALTER TABLE `EmailPT` ADD CONSTRAINT `UniqueEmailPT` UNIQUE(`email`(200));
ALTER TABLE `EmailPT` ADD CONSTRAINT `EmailPT_user_fkey` FOREIGN KEY(`user`) REFERENCES `UserPT`(`id`);
CREATe TABLE `Upsert`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`email` TEXT CHARACTER SET utf8mb4 NOT NULL,`attr` TEXT CHARACTER SET utf8mb4 NOT NULL,`extra` TEXT CHARACTER SET utf8mb4 NOT NULL,`age` BIGINT NOT NULL);
ALTER TABLE `Upsert` ADD CONSTRAINT `UniqueUpsert` UNIQUE(`email`(200));
CREATe TABLE `UpsertBy`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`email` TEXT CHARACTER SET utf8mb4 NOT NULL,`city` TEXT CHARACTER SET utf8mb4 NOT NULL,`attr` TEXT CHARACTER SET utf8mb4 NOT NULL);
ALTER TABLE `UpsertBy` ADD CONSTRAINT `UniqueUpsertBy` UNIQUE(`email`(200));
ALTER TABLE `UpsertBy` ADD CONSTRAINT `UniqueUpsertByCity` UNIQUE(`city`(200));
CREATe TABLE `Strict`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`yes` BIGINT NOT NULL,`no` BIGINT NOT NULL,`def` BIGINT NOT NULL);
test: ConnectionError {errFunction = "query", errNumber = 1005, errMessage = "Can't create table `test`.`Pet` (errno: 150 \"Foreign key constraint is incorrectly formed\")"}
Test suite test: FAIL
Test suite logged to: dist/test/persistent-mysql-2.10.2.3-test.log
0 of 1 test suites (0 of 1 test cases) passed.
```

**Comment by parsonsmatt** [comment]
Please try again with `persistent-2.11`, the FOREIGN syntax got a lot of work with that version.

The `Pet` table migration that is causing the error:

```sql
CREATe TABLE `Pet`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`ownerId` BIGINT NOT NULL REFERENCES `Person`,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`type` TEXT CHARACTER SET utf8mb4 NOT NULL)

-- formatted,
CREATe TABLE `Pet` (
    `id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
    `ownerId` BIGINT NOT NULL REFERENCES `Person`,
    `name` TEXT CHARACTER SET utf8mb4 NOT NULL,
    `type` TEXT CHARACTER SET utf8mb4 NOT NULL
)
```

What immediately looks fishy is that it just says `REFERENCES Person` without a column list. If I toss that into mysql locally, I get this:

```sql
mysql> create table pet2 (id bigint not null auto_increment primary key, owner_id bigint not null references `Person`);
Query OK, 0 rows affected (0.07 sec)

mysql> show create table pet2;
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                   |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------+
| pet2  | CREATE TABLE `pet2` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `owner_id` bigint(20) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Which, uh, doesn't have a foreign key! Even with a column list, it looks like it's just ignored.

```sql
mysql> drop table pet2;
Query OK, 0 rows affected (0.05 sec)

mysql> create table pet2 (id bigint not null auto_increment primary key, owner_id bigint not null references `Person`(`id`));
Query OK, 0 rows affected (0.07 sec)

mysql> show create table pet2;
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                   |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------+
| pet2  | CREATE TABLE `pet2` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `owner_id` bigint(20) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

I think that this is another instance of MySQL parsing something and just ignoring it. Persistent generates an `ALTER TABLE` statemtn:

```sql
ALTER TABLE `Pet` ADD CONSTRAINT `Pet_ownerId_fkey` FOREIGN KEY(`ownerId`) REFERENCES `Person`(`id`);
```

Which should be fine.

With `persistent-2.11`, the migration output I get is:

```sql
CREATe TABLE `Person`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`age` BIGINT NOT NULL,`color` TEXT CHARACTER SET utf8mb4 NULL);
CREATe TABLE `Pet`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`ownerId` BIGINT NOT NULL REFERENCES `Person`  ON DELETE RESTRICT  ON UPDATE RESTRICT,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`type` TEXT CHARACTER SET utf8mb4 NOT NULL);
ALTER TABLE `Pet` ADD CONSTRAINT `Pet_ownerId_fkey` FOREIGN KEY(`ownerId`) REFERENCES `Person`(`id`)  ON DELETE RESTRICT  ON UPDATE RESTRICT;

-- formatted,
CREATe TABLE `Person`(
    `id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
    `name` TEXT CHARACTER SET utf8mb4 NOT NULL,
    `age` BIGINT NOT NULL,
    `color` TEXT CHARACTER SET utf8mb4 NULL
);

CREATe TABLE `Pet`(
    `id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
    `ownerId` BIGINT NOT NULL REFERENCES `Person`  ON DELETE RESTRICT  ON UPDATE RESTRICT,
    `name` TEXT CHARACTER SET utf8mb4 NOT NULL,
    `type` TEXT CHARACTER SET utf8mb4 NOT NULL
);

ALTER TABLE `Pet` 
    ADD CONSTRAINT `Pet_ownerId_fkey` 
    FOREIGN KEY(`ownerId`) REFERENCES `Person`(`id`)  
    ON DELETE RESTRICT  
    ON UPDATE RESTRICT;
```

Which looks fine to me.

**Comment by felixonmars** [comment]
Thanks for the info! I have tried `persistent-2.11` and `persistent-mysql-2.10.3`, the test does proceed further but ended in another error:

```
...
CREATe TABLE `Strict`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`yes` BIGINT NOT NULL,`no` BIGINT NOT NULL,`def` BIGINT NOT NULL);
CREATe TABLE `Relationship`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`name` TEXT CHARACTER SET utf8mb4 NOT NULL,`parent` BIGINT NULL REFERENCES `Relationship` ON DELETE RESTRICT ON UPDATE RESTRICT);
CREATe TABLE `MutA`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`mutB` BIGINT NOT NULL REFERENCES `MutB` ON DELETE RESTRICT ON UPDATE RESTRICT);
ALTER TABLE `MutA` ADD CONSTRAINT `MutA_mutB_fkey` FOREIGN KEY(`mutB`) REFERENCES `MutB`(`id`) ON DELETE RESTRICT ON UPDATE RESTRICT;
CREATe TABLE `MutB`(`id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,`mutA` BIGINT NOT NULL REFERENCES `MutA` ON DELETE RESTRICT ON UPDATE RESTRICT);
ALTER TABLE `MutB` ADD CONSTRAINT `MutB_mutA_fkey` FOREIGN KEY(`mutA`) REFERENCES `MutA`(`id`) ON DELETE RESTRICT ON UPDATE RESTRICT;
test: ConnectionError {errFunction = "query", errNumber = 1005, errMessage = "Can't create table `test`.`Pet` (errno: 150 \"Foreign key constraint is incorrectly formed\")"}
Test suite test: FAIL
```

**Comment by parsonsmatt** [comment]
This must be a discrepancy in how mariadb and mysql parse `CREATE TABLE` statements, since everything works fine with `mysql`.

**Comment by mklinik** [comment]
Persistent creates the foreign key constraint in the Pet table in a two-step process, first CREATe TABLE and then ALTER TABLE. The create table statement contains a half-baked foreign key clause, which MySQL probably silently ignores and MariaDB gives an error message. The whole REFERENCES part in the create table statement can be removed. This is how it should look like, to work with both MySQL and MariaDB:

```SQL
CREATe TABLE `Pet`(
    `id` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
    `ownerId` BIGINT NOT NULL,
    `name` TEXT CHARACTER SET utf8mb4 NOT NULL,
    `type` TEXT CHARACTER SET utf8mb4 NOT NULL
);

ALTER TABLE `Pet` 
    ADD CONSTRAINT `Pet_ownerId_fkey` 
    FOREIGN KEY(`ownerId`) REFERENCES `Person`(`id`)  
    ON DELETE RESTRICT  
    ON UPDATE RESTRICT;
```

See the syntax of foreign key constraints in MySQL and MariaDB.
https://dev.mysql.com/doc/refman/8.0/en/create-table-foreign-keys.html
https://mariadb.com/kb/en/foreign-keys/

I have not tested the above code, this suggestion comes purely from reading the documentation and some hand-written tests in the mariadb console.

**Comment by chrisdone** [comment]
I am also seeing this (slightly formatted version of persistent's "Migrating:" output) for `persistent-2.10.5.2`:

```haskell
Author
    Id Int32
    name Text
    deriving Eq
    deriving Show
Article
    Id Int32
    title Text
    author AuthorId
```
```
CREATe TABLE `author`(
`id` INT(11) NOT NULL PRIMARY KEY,
`name` TEXT CHARACTER SET utf8mb4 NOT NULL
)

CREATe TABLE `article`(`
id` INT(11) NOT NULL PRIMARY KEY,
`title` TEXT CHARACTER SET utf8mb4 NOT NULL,
`author` INT(11) NOT NULL REFERENCES `author`
)
```

MariaDB says:

```
Can't create table `hasura`.`article` (errno: 150 \"Foreign key constraint is incorrectly formed\")
```

Questions:

* Is the persistent-mysql package unstable? Seems like a fairly trivial schema that fails so badly.
* Or should I not be using MariaDB at all with it? 

 I can update to a more recent persistent and switch from MariaDB, if it works... :thinking:

**Comment by chrisdone** [comment]
I've tested with 
```
- persistent-2.13.2.1
- persistent-mysql-2.13.0.2
```

Still seeing the same result, aside from a host of other API changes. I didn't realise persistent was such an unstable library, I thought it'd be unchanging by now. That's my bad, I should have researched more.

I'm going to suggest against using persistent for this client task. We're aiming for simplicity and stability, as little churn as possible.  Sorry to bother.

**Comment by parsonsmatt** [comment]
MariaDB and MySQL have enough syntax differences that I don't really think they can be considered drop-in replacements at this point. IIRC, this isn't the only time that MariaDB users have reported problems with `persistent-mysql` (notably, with things that do not fail on MySQL). `persistent-mysql` is only tested against MySQL, though I'd be happy to accept a PR that adds more explicit MariaDB support (and tests).

`persistent` has been evolving, though I've tried to ensure that each breaking change should be the "last" breaking change in that sense (eg - the `MkSqlBackendArgs` constructor and abstract `SqlBackend` type should mean that adding fields to the `SqlBackend` type for new functionality will only trigger a minor version bump, not a major version bump). If APIs are changing in a manner that breaks your project, I'd be happy to accept an issue - at the least, I can note the failure mode and work to prevent it in the future, and at best, I can provide a work-around and document it better.

**Comment by parsonsmatt** [comment]
Should be fixed in `persistent-mysql-2.13.0.3`.
